Enhanced packaging doc guide (#399)
o3de/o3de-multiplayersample: Documentation/PackedAssetBuilds.md

# Packaged MultiplayerSample Builds

To make relocatable client and server builds for the MultiplayerSample, we recommend making packaged builds. These package builds will contain the Game or Server Launcher and the bundled assets needed to run the launcher outside of the developer environment.

You can make both release packaged builds or profile packaged builds. For more information about creating release builds, see the O3DE documentation on [Creating a Project Game Release Layout for Windows](https://www.o3de.org/docs/user-guide/packaging/windows-release-builds/).

The guide below covers how to make profile packaged builds which are very useful for early sharing and play testing.

## Brief outline of the packaging steps

> 1. Install WWise SDK (required for o3de-multiplayersample but optional for other projects)
> 1. Use the engine from GitHub, not the installer
> 1. Create a build for non-monolithic build of the engine
> 1. Compile the engine with o3de-multiplayersample in profile non-monolithic mode
> 1. Process all the assets using the Asset Processor from profile non-monolithic mode
> 1. Create game bundles (.pak files) from the seed lists (a list of game assets)
> 1. Create a build for monolithic build of the engine
> 1. Compile the engine with o3de-multiplayersample in profile (and/or release) monolithic mode
> 1. Create a folder for the packaged build outside of the engine, this will be the share-able folder of the packaged game
> 1. Copy the monolithic game binaries and the game bundles to the packaged folder
> 1. The packaged build is ready and can be shared without requiring the developer environment.

Reviewer: Possible suggestions to make this list shorter/simpler

```suggestion
1. Install WWise SDK (required for o3de-multiplayersample but optional for other projects)
1. Use the engine from GitHub, not the installer
1. Compile o3de-multiplayersample and the engine (non-monolithic)
1. Process assets using the Asset Processor
1. Create asset bundles (.pak files) from asset seed lists (a list of game assets)
1. Compile o3de-multiplayersample and the engine (monolithic profile or release)
1. Copy the game binaries and asset bundles to a release folder you can share
```

Author: OK

## Pre-requisites

> *Important*: you can not use an installer to produce packaged builds. In order to produce a packaged build you will need to build the engine from source from GitHub.
> 
> Additionally, we will need to compile the engine in non-monolithic profile build and in monolithic profile (and optionally release) mode.
> 
> > Note, a monolithic build is a special build of O3DE engine and projects that combines all the gems used by the project into a single game executable. Additionally, monolithic builds do not build the Editor or the Asset Processor. Thus, a regular non-monolithic build is used to process assets and bundle them into .pak files, while a monolithic build is used to produce share-able game binaries.
> 
> This guide uses `C:\git\o3de` is the source for the O3DE, cloned from GitHub.
> `C:\git\o3de-multiplayersample` is the location of the clone project. These paths are optional. If you choose to use different paths, amend the command scripts in the rest of the guide accordingly.

Reviewer: ```suggestion

1. Your computer is [setup to build O3DE](https://www.o3de.org/docs/user-guide/build/)
1. You have the O3DE engine source code from GitHub [https://github.com/o3de/o3de](https://github.com/o3de/o3de)
 
*Important*: you can not use an O3DE SDK installer to produce packaged builds, that is why you need the engine source code from GitHub.

NOTE: This guide uses `C:\git\o3de` as the location where the O3DE engine source code has been cloned from GitHub.
`C:\git\o3de-multiplayersample` is the location of where the project has been cloned from GitHub. These paths are optional. If you use different paths, amend the commands in the rest of the guide accordingly.
```

Author: The reason I put it this way is because I want to call this requirement out early in the document and then re-iterate later on.

## Windows Profile PAK Setup

> Multiplayer Sample uses WWise gem and assets for audio effects. O3DE engine (and O3DE installers) do not include WWise support by default. In order to add the WWise support in the engine, one must first install WWise SDK and then re-build O3DE engine from source.

Reviewer: I'm pretty sure the proper spelling is `Wwise` and not `WWise`  https://www.audiokinetic.com/en/products/wwise/
```suggestion
Multiplayer Sample uses the Wwise gem for audio effects and so we must first install the Wwise SDK.
```

![Wwise version selection](Media/Wwise_installer_version_selection.png)

> REBOOT (or logout / login). Otherwise, the environment settings won't get picked up for any builds in Visual Studio. They will only apply to command-line builds, and only for any command-line windows that have been opened after the installer finishes.

### Build o3de-multiplayersample and the engine (non-monolithic)

Build a regular profile build of the game as per the [README.md](../README.md) in an engine-centric way. Be sure to build from the source engine and not the installer. Here is an example:

Clone the engine from source.

```shell
C:\git> git clone https://github.com/o3de/o3de
```

Navigate to `C:\git\o3de`.
Create a build folder for non-monolithic build of the engine, for example `C:\git\o3de\build\non_mono`.

> The location and the name of the build folder is optional.

Configure the engine in a non-monolithic mode with o3de-multiplayersample project.

```shell
C:\git\o3de> cmake -S . -B build\non_mono -DLY_MONOLITHIC_GAME=OFF -DLY_PROJECTS="C:\git\o3de-multiplayersample"
```

> `-DLY_MONOLITHIC_GAME=OFF` is the default value but for clarity it's specified here explicitly.

> Build the Editor. This will compile the project and necessary gems to produce the required game assets.

Reviewer: If you choose to include the AssetBundler in this build command, you could say `Build the Editor and AssetBundler`

Author: Done

```shell
C:\git\o3de> cmake --build build\non_mono --target Editor --config profile
```

Run the Asset Processor with o3de-multiplayersample and let all the assets get processed.

```shell
C:\git\o3de> .\build\non_mono\bin\profile\AssetProcessor.exe --project-path C:\git\o3de-multiplayersample
```

### Build AssetBuilder

You will need to build the [AssetBundler](https://www.o3de.org/docs/user-guide/packaging/asset-bundler/overview/) tool if not built to create game bundles (.pak files).

For example:
```shell

> C:\git\o3de> cmake --build build\non_mono --target AssetBundler --config profile

Reviewer: you could combine this with the Editor build command above by adding the `AssetBundler` target after `Editor`

```

### Build monolithic game

Build a second version of the executables as monolithic pak builds.

> A separate build folder is required for building monolithic binaries, separate from the non-monolithic build folder.

```shell
# Create build files for a monolithic build that also disables all user/project registry settings overrides

> C:\git\o3de> cmake -S . -B build\mono -DLY_MONOLITHIC_GAME=1 -DALLOW_SETTINGS_REGISTRY_DEVELOPMENT_OVERRIDES=0 -DLY_PROJECTS="C:\git\o3de-multiplayersample"

Reviewer: it's a little odd to me to use `OFF` in the statement above for this parameter and `1` here instead of `ON` but I'm guessing our docs use `1`

Author: Changed to zero

# Build the profile versions of all the executables
C:\git\o3de> cmake --build build\mono --target MultiplayerSample.GameLauncher MultiplayerSample.ServerLauncher MultiplayerSample.UnifiedLauncher --config profile
```

Profile monolithic game binaries will be located in `C:\git\o3de\build\mono\bin\profile`.

> Optionally, you can build monolithic release game binaries.

Reviewer: I wonder if we should have a warning here about how if they build in release mode the build won't have access to the console (unless the cvar is enabled) and there may be fewer logs etc.

You can make both release packaged builds or profile packaged builds. For more information about creating release builds, see the O3DE documentation on [Creating a Project Game Release Layout for Windows](https://www.o3de.org/docs/user-guide/packaging/windows-release-builds/).

The guide below covers how to make profile packaged builds which are very useful for early sharing and play testing.

## Brief outline of the packaging steps

1. Install Wwise SDK (required for o3de-multiplayersample but optional for other projects)
1. Use the engine from GitHub, not the installer
1. Compile o3de-multiplayersample and the engine (non-monolithic)
1. Process assets using the Asset Processor
1. Create asset bundles (.pak files) from asset seed lists (a list of game assets)
1. Compile o3de-multiplayersample and the engine (monolithic profile or release)
1. Copy the game binaries and asset bundles to a release folder you can share

## Pre-requisites

> *Important*: you can not use an installer to produce packaged builds. In order to produce a packaged build you will need to build the engine from source from GitHub.

Reviewer 2: ```suggestion
*Important*: you can not use an installer to produce packaged builds. In order to produce a packaged build, you will need to build the engine from source from GitHub.
```

Author: OK

![Wwise version selection](Media/Wwise_installer_version_selection.png)

> REBOOT (or logout / login). Otherwise, the environment settings won't get picked up for any builds in Visual Studio. They will only apply to command-line builds, and only for any command-line windows that have been opened after the installer finishes.

### Build o3de-multiplayersample and the engine (non-monolithic)

Build a regular profile build of the game as per the [README.md](../README.md) in an engine-centric way. Be sure to build from the source engine and not the installer. Here is an example:

Clone the engine from source.

```shell
C:\git> git clone https://github.com/o3de/o3de
```

Navigate to `C:\git\o3de`.
Create a build folder for non-monolithic build of the engine, for example `C:\git\o3de\build_non_mono`.

```shell
C:\git\o3de> mkdir build_non_mono
```
> The location and the name of the build folder is optional.

Configure the engine in a non-monolithic mode with o3de-multiplayersample project.

```shell
C:\git\o3de> cmake -S . -B build_non_mono -DLY_MONOLITHIC_GAME=OFF -DLY_PROJECTS="C:\git\o3de-multiplayersample"

> ```

Reviewer 3: CMake will create the build folder for the user.
So the step to make the directory is not needed.

Also I recommend creating the build folder inside of the `<engine-root>/build` directory as that directory is part of the .gitignore file and will not show up as modifications to the repo.

Author: Sounds good, updated

> `-DLY_MONOLITHIC_GAME=OFF` is the default value but for clarity it's specified here explicitly.

Build the Editor. This will compile the project and necessary gems to produce the required game assets.

```shell
C:\git\o3de> cmake --build build_non_mono --target Editor --config profile
```

Run the Asset Processor with o3de-multiplayersample and let all the assets get processed.

```shell
C:\git\o3de> .\build_non_mono\bin\profile\AssetProcessor.exe --project-path C:\git\o3de-multiplayersample
```

### Build AssetBuilder

You will need to build the [AssetBundler](https://www.o3de.org/docs/user-guide/packaging/asset-bundler/overview/) tool if not built to create game bundles (.pak files).

For example:
```shell
C:\git\o3de> cmake --build build_non_mono --target AssetBundler --config profile
```

### Build monolithic game

Build a second version of the executables as monolithic pak builds.

> A separate build folder is required for building monolithic binaries, separate from the non-monolithic build folder.

```shell
C:\git\o3de> mkdir build_mono

# Create build files for a monolithic build that also disables all user/project registry settings overrides

> C:\git\o3de> cmake -S .. -B build_mono -DLY_MONOLITHIC_GAME=1 -DALLOW_SETTINGS_REGISTRY_DEVELOPMENT_OVERRIDES=0 -DLY_PROJECTS="C:\git\o3de-multiplayersample"

Reviewer 2: Just wanted to confirm you meant to write ".." here?

Author: You are correct. It should be a single dot.

> REBOOT (or logout / login). Otherwise, the environment settings won't get picked up for any builds in Visual Studio. They will only apply to command-line builds, and only for any command-line windows that have been opened after the installer finishes.

> ### Build non-monolithic profile builds of the Engine

Reviewer: ```suggestion
### Build o3de-multiplayersample and the engine (non-monolithic)
```

Build a regular profile build of the game as per the [README.md](../README.md) in an engine-centric way. Be sure to build from the source engine and not the installer. Here is an example:

Clone the engine from source.

```shell
C:\git> git clone https://github.com/o3de/o3de
```

Navigate to `C:\git\o3de`.
Create a build folder for non-monolithic build of the engine, for example `C:\git\o3de\build_non_mono`.

```shell

> C:\git\o3de> mkdir build_non_mono
> ```
> > The location and the name of the build folder is optional.
> 
> Configure the engine in a non-monolithic mode with o3de-multiplayersample project.
> 
> ```shell
> C:\git\o3de\build_non_mono> cmake .. -DLY_MONOLITHIC_GAME=OFF -DLY_PROJECTS="C:\git\o3de-multiplayersample"

Reviewer 3: Well this is missing the change directory call to build_non_mono.
Though I see the old steps as superior as it uses the -B and -S option to avoid the need of additional change directroy call.

```suggestion
C:\git\o3de> mkdir build_non_mono
```
> The location and the name of the build folder is optional.

Configure the engine in a non-monolithic mode with o3de-multiplayersample project.

```shell
C:\git\o3de\> cmake -B build_non_mono -S. -DLY_MONOLITHIC_GAME=OFF -DLY_PROJECTS="C:\git\o3de-multiplayersample"
```

Author: OK

> C:\git\o3de\build_non_mono> cmake .. -DLY_MONOLITHIC_GAME=OFF -DLY_PROJECTS="C:\git\o3de-multiplayersample"

Reviewer: Are you using engine centric to avoid having to run `get_python.bat`, register the engine and the project?

I kinda wish you could keep the project outside the engine folder to avoid git thinking there are a ton of changes
```shell
# get the engine and python
git clone https://github.com/o3de/o3de  c:\git\o3de
c:\git\o3de\python\get_python.bat

# get the project and tell it to use the engine
git clone https://github.com/o3de/o3de-multiplayersample c:\git\o3de-multiplayersample
c:\git\o3de\scripts\o3de.bat edit-project-properties -pp c:\git\o3de-multiplayersample --engine-path c:\git\o3de --user

# configure and build profile non-mono
cd c:\git\o3de-multiplayersample
cmake --preset windows-default
cmake --build build/windows --target Editor AssetBundler --config profile

# configure and build mono profile
cmake --preset  windows-mono-default  -DALLOW_SETTINGS_REGISTRY_DEVELOPMENT_OVERRIDES=0 -DLY_PROJECTS="C:\git\o3de-multiplayersample"
cmake --build build/windows_mono --target MultiplayerSample.GameLauncher MultiplayerSample.ServerLauncher MultiplayerSample.UnifiedLauncher --config profile

# Process assets
c:\git\o3de-multiplayersample\build\windows\bin\profile\AssetProfilerBatch.exe

# Bundle assets - there's probably a way to pass in the seed lists from the command line
c:\git\o3de-multiplayersample\build\windows\bin\profile\AssetBundlerBatch.exe --foo=path/to/seedlist --bundle=path/to/engine.pak
c:\git\o3de-multiplayersample\build\windows\bin\profile\AssetBundlerBatch.exe --foo=path/to/seedlist --bundle=path/to/game.pak

# Copy binaries and assets into release folder
cp blah blah blah
```

Author: The project is outside of the engine: C:\git\o3de\build_non_mono versus C:\git\o3de-multiplayersample. But I am keeping engine-centric otherwise, you have to add steps on registering a new install layout to the steps, which would add a lot of complexity to these steps.

```

> `-DLY_MONOLITHIC_GAME=OFF` is the default value but for clarity it's specified here explicitly.

Build the Editor. This will compile the project and necessary gems to produce the required game assets.

```shell
C:\git\o3de\build_non_mono> cmake --build . --target Editor --config profile
```

Run the Asset Processor with o3de-multiplayersample and let all the assets get processed.

```shell
C:\git\o3de\build_non_mono> .\bin\profile\AssetProcessor.exe --project-path C:\git\o3de-multiplayersample
```

### Build AssetBuilder

You will need to build the [AssetBundler](https://www.o3de.org/docs/user-guide/packaging/asset-bundler/overview/) tool if not built to create game bundles (.pak files).

For example:
```shell
C:\git\o3de\build_non_mono> cmake --build . --target AssetBundler --config profile
```

### Test the profile build

* Open the game in editor
    * load `NewStarBase` level
    * Verify that game can launch and connect to local server from editor
* Validate local game launcher can connect to local server

> 

Reviewer: I recommend putting the ideal steps at the top and move the testing to a `Troubleshooting` section at the bottom.  If all a user wants to do is package MPS they should just be able to follow the list at the top, but if something goes wrong they can look further down the page

Author: I just removed this section. The main readme already covers this.
